core/src/main/java/org/apache/iceberg/variants/VariantMetadata.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.variants;
+
+/** A variant metadata dictionary. */
+public interface VariantMetadata extends Variants.Serialized {
+  /** Returns the ID for a {@code name} in the dictionary, or -1 if not present. */
+  int id(String name);
+
+  /**
+   * Returns the field name for an ID in metadata.
+   *
+   * @throws ArrayIndexOutOfBoundsException if the dictionary does not contain the ID
```

Comment:
This doesn't feel like the right exception to throw here.  It's misleading that it has to do with the size of the array.

Reply:
I'm not sure about `NoSuchElementException`. That's thrown when `Iterator#next` is called on an empty iterator.

I like `ArrayIndexOutOfBoundsException` because the metadata is basically an array of strings and the IDs are the indexes of the array.